src/IL/MethodBodyWriter.cs

```diff
@@ -100,7 +100,7 @@ public static void Write(IAssemblyMetadata metadata, IReadOnlyList<Instruction>
                         switch (il[i].Operand)
                         {
                             case Type type:
-                                metadata.ILBuilder.WriteInt32(MetadataTokens.GetToken(metadata.GetTypeHandle(type)));
+                                metadata.ILBuilder.WriteInt32(MetadataTokens.GetToken(metadata.GetTypeHandle(type, true)));
```

Comment:
I would suggest to write `GetTypeHandle(type, inMethodBodyWritingContext: true)`. This is a bit redundant, but as it's an edge-case, it makes sense to locally clarify the intent.